### PR TITLE
six set of modules

### PR DESCRIPTION
**Pull request by carchi8py**
##### SUMMARY
six set of 5
We are updating and adding the Netapp ONTAP specific modules. The update modules add new options for the existing functions. The new modules allow for setting up and modify different part of the ONTAP system.

We have about 33 modules, we've been told to start with 2 or 3 first, and take feedback on these first before pushing them all out

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
	na_ontap_snmp.py
	na_ontap_svm.py
	na_ontap_ucadapter.py
	na_ontap_user.py
	na_ontap_user_role.py

##### ANSIBLE VERSION

```
ansible 2.6.0 (devel 0d1e9bfe90) last updated 2018/04/23 14:26:54 (GMT -400)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /tmp/ansible/lib/ansible
  executable location = /tmp/ansible/bin/ansible
  python version = 2.7.5 (default, Oct 11 2015, 17:47:16) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by schmots1** [comment]
shipit

**Comment by hulquest** [comment]
shipit

File: lib/ansible/modules/storage/netapp/na_ontap_snmp.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    def delete_snmp_community(self):
+        """
+        Delete a SNMP community
+        """
+        return self.invoke_snmp_community('snmp-community-delete')
+
+    def apply(self):
+        """
+        Apply action to SNMP community
+        We want the module to be idempotent
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_snmp.py]: er, but it's not (at least it doesn't report as such). Is the issue that there's no way to enumerate existing communities?

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+  name:
+    description:
+    - The name of the SVM to manage.
+    required: true
+
+  new_name:
+    description:
+    - New name of the SVM to be renamed
+    required: false
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py]: redundant

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    - Required when C(state=present).
+
+  root_volume_security_style:
+    description:
+    -   Security Style of the root volume.
+    -   When specified as part of the vserver-create,
+        this field represents the security style for the Vserver root volume.
+    -   When specified as part of vserver-get-iter call,
+        this will return the list of matching Vservers.
+    -   Possible values are 'unix', 'ntfs', 'mixed'.
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py]: redundant

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    - When specified as part of a vserver-create,
+      this field represent the list of protocols allowed on the Vserver.
+    - When part of vserver-get-iter call,
+      this will return the list of Vservers
+      which have any of the protocols specified
+      as part of the allowed-protocols.
+    - When part of vserver-modify,
+      this field should include the existing list
+      along with new protocol list to be added to prevent data disruptions.
+    - Possible values
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:79]: use `choices`

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+  allowed_protocols:
+    description:
+    - Allowed Protocols.
+    - When specified as part of a vserver-create,
+      this field represent the list of protocols allowed on the Vserver.
+    - When part of vserver-get-iter call,
+      this will return the list of Vservers
+      which have any of the protocols specified
+      as part of the allowed-protocols.
+    - When part of vserver-modify,
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:76]: this should probably be reworded without using API-isms- it's basically the thing discussed before on other list types, where you're accepting the absolute list, not a set to be merged with the existing list.

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    - nvme   NVMe protocol
+
+  aggr_list:
+    description:
+    - List of aggregates assigned for volume operations.
+    - These aggregates could be shared for use with other Vservers.
+    - When specified as part of a vserver-create,
+      this field represents the list of aggregates
+      that are assigned to the Vserver for volume operations.
+    - When part of vserver-get-iter call,
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:95]: APIism

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    - iscsi   iSCSI protocol,
+    - ndmp   NDMP protocol,
+    - http   HTTP protocol,
+    - nvme   NVMe protocol
+
+  aggr_list:
+    description:
+    - List of aggregates assigned for volume operations.
+    - These aggregates could be shared for use with other Vservers.
+    - When specified as part of a vserver-create,
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:92]: APIism

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+            name=dict(required=True, type='str'),
+            new_name=dict(required=False, type='str'),
+            root_volume=dict(type='str'),
+            root_volume_aggregate=dict(type='str'),
+            root_volume_security_style=dict(type='str', choices=['unix',
+                                                                 'ntfs',
+                                                                 'mixed',
+                                                                 'unified'
+                                                                 ]),
+            allowed_protocols=dict(type='list'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:143]: choices

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+
+module: na_ontap_svm
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:18]: It seems like "SVM" is the preferred nomenclature for this- why is `vserver` used everywhere in the module args and docs?

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+
+        rename_vserver = False
+        modify_protocols = False
+        modify_aggr_list = False
+        obj = open('vserver-log', 'a')
+        if vserver_details is not None:
+            if self.state == 'absent':
+                changed = True
+            elif self.state == 'present':
+                if self.new_name is not None and self.new_name != self.name:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:303]: rename is not idempotent, as discussed previously

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+                    vserver_details['allowed_protocols'].sort()
+                    if self.allowed_protocols != vserver_details['allowed_protocols']:
+                        modify_protocols = True
+                        changed = True
+                if self.aggr_list is not None:
+                    self.aggr_list.sort()
+                    vserver_details['aggr_list'].sort()
+                    if self.aggr_list != vserver_details['aggr_list']:
+                        modify_aggr_list = True
+                        changed = True
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_svm.py:317]: what about root_volume_X args?

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    - netapp.na_ontap
+version_added: '2.6'
+author: chhaya gunawat (chhayag@netapp.com)
+
+description:
+    - modify the UC adapter mode and type taking pending type and mode into account.
+
+options:
+  state:
+    description:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py:31]: why does this arg exist?

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+  node_name:
+    description:
+    - Specifies the adapter home node.
+    required: true
+
+  mode:
+    description:
+    - Specifies the mode of the adapter.
+    required: false
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py]: redundant

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    description:
+    - Specifies the adapter name.
+    required: true
+
+  node_name:
+    description:
+    - Specifies the adapter home node.
+    required: true
+
+  mode:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py:47]: choices?

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+    description:
+    - Specifies the mode of the adapter.
+    required: false
+
+  type:
+    description:
+    - Specifies the fc4 type of the adapter.
+    required: false
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py]: redundant

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+    description:
+    - Specifies the adapter home node.
+    required: true
+
+  mode:
+    description:
+    - Specifies the mode of the adapter.
+    required: false
+
+  type:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py]: choices?

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+    ''' object to describe  adapter info '''
+
+    def __init__(self):
+
+        self.argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=['present'], default='present'),
+            adapter_name=dict(required=True, type='str'),
+            node_name=dict(required=True, type='str'),
+            mode=dict(required=False, type='str'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py:93]: choices

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+            type=dict(required=False, type='str'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py:94]: choices

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+        def need_to_change(expected, pending, current):
+            if expected is None:
+                return False
+            if pending is not None:
+                return pending != expected
+            if current is not None:
+                return current != expected
+            return False
+
+        if adapter_detail:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py:177]: failure else case?

File: lib/ansible/modules/storage/netapp/na_ontap_user.py [hunk 20]
```diff
@@ -0,0 +1,392 @@
...
+
+description:
+- Create or destroy users.
+
+options:
+
+  state:
+    description:
+    - Whether the specified user should exist or not.
+    required: true
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user.py]: incorrect with default specified; remove

File: lib/ansible/modules/storage/netapp/na_ontap_user.py [hunk 21]
```diff
@@ -0,0 +1,392 @@
...
+    choices: ['present', 'absent']
+    default: 'present'
+
+  name:
+    description:
+    - The name of the user to manage.
+    required: true
+
+  application:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user.py:41]: should be `applications` to imply list?

File: lib/ansible/modules/storage/netapp/na_ontap_user.py [hunk 22]
```diff
@@ -0,0 +1,392 @@
...
+    - For an existing user, this value will be used as the new password.
+
+  role_name:
+    description:
+    - The name of the role. Required when C(state=present)
+
+  lock_user:
+    description:
+    - To ock/unlock user.
+    required: false
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user.py]: redundant

File: lib/ansible/modules/storage/netapp/na_ontap_user.py [hunk 23]
```diff
@@ -0,0 +1,392 @@
...
+    - It is ignored for creating snmp users, but is required for creating non-snmp users.
+    - For an existing user, this value will be used as the new password.
+
+  role_name:
+    description:
+    - The name of the role. Required when C(state=present)
+
+  lock_user:
+    description:
+    - To ock/unlock user.
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user.py]: "Whether the specified user is locked"

File: lib/ansible/modules/storage/netapp/na_ontap_user.py [hunk 24]
```diff
@@ -0,0 +1,392 @@
...
+
+  name:
+    description:
+    - The name of the user to manage.
+    required: true
+
+  application:
+    description:
+    - Applications to grant access to.
+    required: true
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user.py]: should only be required when state is present?

File: lib/ansible/modules/storage/netapp/na_ontap_user.py [hunk 25]
```diff
@@ -0,0 +1,392 @@
...
+    - Password for console application
+    - Password, domain, nsswitch, cert for http application.
+    - Password, domain, nsswitch, cert for ontapi application.
+    - Community for snmp application (when creating SNMPv1 and SNMPv2 users).
+    - The usm and community for snmp application (when creating SNMPv3 users).
+    - Password for sp application.
+    - Password for rsh application.
+    - Password for telnet application.
+    - Password, publickey, domain, nsswitch for ssh application.
+    required: true
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user.py:61]: should only be required when state is present?

File: lib/ansible/modules/storage/netapp/na_ontap_user_role.py [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+
+description:
+- Create or destroy user roles
+
+options:
+
+  state:
+    description:
+    - Whether the specified user should exist or not.
+    required: true
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user_role.py]: incorrect with `default` specified; remove

File: lib/ansible/modules/storage/netapp/na_ontap_user_role.py [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+
+  name:
+    description:
+    - The name of the role to manage.
+    required: true
+
+  command_directory_name:
+    description:
+    - The command or command directory to which the role has an access.
+    required: true
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_user_role.py:44]: only required when state==present?

**Review by nitzmahone — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/storage/netapp/na_ontap_snmp.py (hunk 1 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_snmp.py]: We will look at this in the next release.

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py (hunk 10 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_svm.py:303]: We'll address this in the next release.

File: lib/ansible/modules/storage/netapp/na_ontap_svm.py (hunk 11 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_svm.py:317]: We'll look in to this in the next release.

File: lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py (hunk 19 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_ucadapter.py:177]: We'll look in to this in the next release.

File: lib/ansible/modules/storage/netapp/na_ontap_user.py (hunk 21 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_user.py:41]: We'll look at making a list for the next release. Fixed the comment to be singular.

File: lib/ansible/modules/storage/netapp/na_ontap_user.py (hunk 23 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_user.py]: fixed

File: lib/ansible/modules/storage/netapp/na_ontap_user.py (hunk 24 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_user.py]: We'll look in to this for the next release.

File: lib/ansible/modules/storage/netapp/na_ontap_user.py (hunk 25 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_user.py:61]: We'll look in to this for the next release.

File: lib/ansible/modules/storage/netapp/na_ontap_user_role.py (hunk 27 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_user_role.py:44]: We'll look in to this for the next release.

**Pull request merged**